text/0000-min-target-api-version.md

```diff
@@ -0,0 +1,205 @@
+
+By default, the standard library uses a sensible minimum API version. For example, for `x86_64-pc-windows-msvc` the minimum API version is "6.1.7600" which corresponds to Windows 7. However, there's good reason why you might want control of how your code is compiled depending on what the minimum API version is set as. For instance, if you want to:
+
+* set your crate's minimum API version higher than that of the standard library. 
+* change certain implementation details of your crate depending on what a downstream user sets their minimum API version to be.  
+* have some sensible compiler error if users of your crate require a lower minimum API version that you require.
+
+This is where certain `cfg` predicates come in that allow you to specify what the minimum target APIs are. These include the following:
+* `min_windows_build_version`: the [minimum build version](https://gaijin.at/en/infos/windows-version-numbers) of Windows.
+* `min_libc_version`: the version of libc in use.
+* `min_kernel_version`: the version of the kernel in use.
+* `min_macos_version`: The minium version of macOS in use.
+
+
+Let's use `min_windows_build_number` as an example. It should be clear how this example would be extended to the other `cfg` predicates. The `min_windows_build_number` allows you to conditionally compile code based on the set minimum API version. For example an implementation of mutex locking on Windows might look like this:
+
+```rust
+pub unsafe fn unlock(&self) {
+    *self.held.get() = false;
+    if cfg!(min_windows_build_version >= "6.0.6000") { // API version greater than Vista
+        c::ReleaseSRWLockExclusive(raw(self)) // Use the optimized ReleaseSRWLockExclusive routine
+    } else {
+        (*self.remutex()).unlock()  // Fall back to an alternative that works on older Windows versions
+    }
+}
+```
+
+End users can set the `min_windows_build_version` in the Cargo configuration file `.cargo/config` under the [`target key`](https://doc.rust-lang.org/cargo/reference/config.html#target) like so:
+
+```toml
+[target.x86_64-pc-windows-msvc]
+min_windows_build_version = "6.0.6000" # Vista 
+```
+
+When using `rustc` directly, this option can be provided as a command line option: `rustc --cfg 'min_windows_build_version="6.0.6000"'`.
+
+If an end user sets their `min_windows_build_version` to an incompatible version then the user receives an error. For instance, in the example above where the user is setting their `min_windows_build_version` to Windows Vista, they will receive an error when linking with the standard library which imposes Windows 7 as its `min_windows_build_version` by default for the `x86_64-pc-windows-msvc` target.
+
+If a library does not explicitly set its `min_windows_build_version`, it will automatically be set to the largest `min_windows_build_version` of all of its transitive dependencies.
```

Comment:
How would a library set its minimum version? I don't see anything to that effect in this section nor in the reference-level explanation.

Reply:
This is actually an unanswered question - thanks for bringing it up. The RFC mentions storing this information in the crate metadata but gives no mechanism for doing so. I'm not aware of any existing mechanism that we could use for such a thing. The closest that comes to mind is `#![no_std]`, but having a `#![min_windows_build_version]` seems like overkill. 

If we switch to a dedicated syntax like `#[min_version(...)]` we could conceivably just store the smallest of those used. Perhaps this is another point in favor of the dedicated syntax.